### DISCUSSION TITLE
Error when using dataframe

### DISCUSSION THREAD

**Discussion by michail-p** [microsoft/FLAML#289] [Q&A]
Im getting this error https://pastebin.com/3VS3AdXg when i run this code https://pastebin.com/1xrJDqrc any ideas? the csv contains 22 columns, first named 'target' which is the label and the rest are the features

**Comment by sonichi** [comment]
From the error message the problem looks related to the feature types. Could you share a minimal dataframe that causes this error?

**Reply by michail-p** [reply]
That was it, i changed the data structure and it worked but now im getting an error when i try to load the model using pickle.
https://pastebin.com/VTZpreQN

the code is this 

```
        with open("model.pkl", "rb") as model_file:
            model = pickle.load(model_file)
```

**Reply by sonichi** [reply]
I can't reproduce this error in Python 3.9. Could you share a reproducible example and the environment you are using?

**Reply by michail-p** [reply]
I was able to fix that too, i hat to use join() to my threads. I have a problem with the prediction feed. i get this error when i predict: ValueError: Number of features of the model must match the input. Model n_features_ is 63 and input n_features is 1

For reference i train the model using this type of data {'label': 'xxx', 'feature1': 'xxx' ... }
and the prediction input is a list of decimals [1, 2, 3, 4 ...] same length as the training features

**Reply by sonichi** [reply]
Are the data in a dictionary? Is it possible to convert it to either numpy array or pandas dataframe, which is required as the type of X_train?

**Reply by michail-p** [reply]
I did but the prediction results are weird. It only returns one label and when i print the numpy array i looks weird too. Using reshape didnt work. Ill try as dataframe

**Reply by sonichi** [reply]
It sounds that the data are not prepared properly. The numpy array for X_train should be a 2D array with m columns as m features. Could you print a few rows of the array?

**Reply by michail-p** [reply]
Here you can see the training dataframe https://docs.google.com/spreadsheets/d/12GX0J9TPWate7YpEcsQhmS8hpL2KAzIQBSh-mYbg0nY/edit?usp=sharing

And this is the input for prediction 
https://pastebin.com/LcmTJXCm

**Reply by sonichi** [reply]
The training data look all right. For the input for prediction, is it a 2D array? What I see is
['0.25122037529945373535' '0.74068242311477661133'
 '0.00000006014916920094' '0.30432972311973571777'
 '0.69939041137695312500' '-0.02410023845732212067'
 '0.33621916174888610840' '0.62934952974319458008'
 '-0.03792585060000419617' '0.33578974008560180664'
 '0.55962210893630981445' '-0.05288818851113319397'
 '0.32608222961425781250' '0.50164848566055297852'
 '-0.06819321215152740479' '0.31425037980079650879'
 '0.53874665498733520508' '-0.02467181533575057983'
 '0.33330190181732177734' '0.43351542949676513672'
 '-0.05031802132725715637' '0.34556096792221069336'
 '0.37221688032150268555' '-0.06602224707603454590'
 '0.35269880294799804688' '0.31813460588455200195'
 '-0.07665584981441497803' '0.27828100323677062988'
 '0.54176425933837890625' '-0.02772625908255577087'
 '0.29660168290138244629' '0.48881703615188598633'
 '-0.07290106266736984253' '0.30656087398529052734'
 '0.55461257696151733398' '-0.08532354980707168579'
 '0.30652394890785217285' '0.60849010944366455078'
 '-0.08397462964057922363' '0.24166814982891082764'
 '0.55640423297882080078' '-0.03328468650579452515'
 '0.25836661458015441895' '0.49788558483123779297'
 '-0.07529371976852416992' '0.27569353580474853516'
 '0.56327807903289794922' '-0.06782094389200210571'
 '0.28025814890861511230' '0.61756694316864013672'
 '-0.05201809853315353394' '0.20187914371490478516'
 '0.58179593086242675781' '-0.04080755263566970825'
 '0.21699039638042449951' '0.53692662715911865234'
 '-0.07093623280525207520' '0.24065217375755310059'
 '0.57719737291336059570' '-0.06498146802186965942'
 '0.25063088536262512207' '0.61693632602691650391'
 '-0.05276469886302947998']

**Reply by michail-p** [reply]
Thats how it prints. Its supposed to be 1D and each value is a feature. This is the code https://pastebin.com/rKJ1qssv

**Reply by sonichi** [reply]
The predict() function requires a 2D array as input. If you input is a single instance, please format it in shape 1 * m (m=#features).

**Reply by michail-p** [reply]
It worked! I trained it for an hour and can exceptionally recognize two hand gestures :D

**Reply by sonichi** [reply]
Sounds great. It probably doesn't require one hour. You can check the console output about when the best model is found in the end.

**Reply by michail-p** [reply]
Is there a way to make it stop automatically when the best model is found?

And is there a way to calculate how many samples i need for each label?

**Reply by sonichi** [reply]
> Is there a way to make it stop automatically when the best model is found?

Before the search is over, one cannot tell whether the current best model will be the best model in the end. What I suggested is a postmortem analysis. You can use `early_stop="True"`, which uses a conservative condition to stop the search, i.e., the total search time is 10 times of the time taken to find the currently best model. Also refer to https://github.com/microsoft/FLAML/wiki/Time-budget.
> 
> And is there a way to calculate how many samples i need for each label?

I need to understand this question. Do you mean you want to incrementally increase the number of labeled training examples until the accuracy is above a threshold?

**Reply by michail-p** [reply]
Yes but i dont know how many samples to start with

i also get ConvergenceWarning: The max_iter was reached which means the coef_ did not converge sometimes, what does it mean?

**Reply by sonichi** [reply]
You can start with a small number, say 100, and then apply the IterTrain algorithm as in this [paper](http://www.vldb.org/pvldb/vol13/p2215-dutt.pdf). You need to replace the confidence interval calculation part because your task is classification. For computing confidence interval for classification accuracy, refer to this [paper](http://export.arxiv.org/abs/1811.03250).

The warning is for the estimator lrl1 (logistic regression with l1 regularization). What's the best estimator selected eventually?

**Reply by michail-p** [reply]
Thanks Chi but theres no way im going through those papers i need some ready solution thats why i picked flaml 😂

The estimator chosen was extra tree. I also used early stop and it stopped at 48 mins instead 60 which was the time i put. Weird thing is before i use early stop it picked another estimator.

**Reply by sonichi** [reply]
Thanks. As far as I know these techniques are not present in any existing library. I've added your feature request to https://github.com/microsoft/FLAML/wiki/Roadmap-for-Upcoming-Features. I wish someone could pick it up. In the same time, the more you could share about your use case and its significance, the more likely it will be prioritized.

**Reply by michail-p** [reply]
Thanks, should i just write about it here or somewhere specific?

**Reply by sonichi** [reply]
How about creating a new thread?

**Reply by michail-p** [reply]
Thanks done
